src/rebar3_hex_file.erl

```diff
@@ -1,34 +1,100 @@
 -module(rebar3_hex_file).
 
--export([ expand_paths/2
-        , update_app_src/2
-        ]).
+-include_lib("kernel/include/file.hrl").
 
-expand_paths(Paths, Dir) ->
-    AbsDir = filename:absname(Dir),
-    Files = lists:flatmap(fun dir_files1/1, [filename:join(Dir, P) || P <- Paths]),
-    [{relative_path(F1, AbsDir), F1} || F1 <- filter_regular(Files)].
+-export([
+    expand_paths/2,
+    update_app_src/2
+]).
 
-dir_files1(Dir) ->
-    lists:flatmap(fun(Y) -> dir_files(Y) end, filelib:wildcard(Dir)).
+expand_paths(Paths, Dir) ->
+    AbsDir = filename:absname(rebar_utils:to_binary(Dir)),
+    Files1 = [filename:join(AbsDir, rebar_utils:to_binary(P)) || P <- Paths],
+    Files2 = lists:flatmap(
+        fun(F) ->
+            [rebar_utils:to_binary(F1) || F1 <- filelib:wildcard(rebar_utils:to_list(F))]
+        end,
+        Files1
+    ),
+    Files3 = lists:flatmap(fun dir_files/1, Files2),
+    Files4 = lists:map(fun expand/1, Files3),
+    Files5 = maps:keys(lists:foldl(fun(K, Acc) -> Acc#{K => true} end, #{}, Files4)),
+    case Files5 of
+        [] ->
+            [];
+        _ ->
+            [
+                {rebar_utils:to_list(relative_to(F1, AbsDir)), rebar_utils:to_list(F1)}
+             || F1 <- filter_regular(Files5)
+            ]
+    end.
 
 filter_regular(Files) ->
     lists:filter(fun filelib:is_regular/1, [filename:absname(F) || F <- Files]).
 
 dir_files(Path) ->
-    case filelib:is_dir(Path) of
-        true ->
-            filelib:wildcard(filename:join(Path, "**"));
-        false ->
+    case file:read_link_info(Path, [{time, universal}]) of
+        {ok, FileInfo} ->
+            case FileInfo#file_info.type of
+                directory ->
+                    {ok, More} = file:list_dir(Path),
+                    Paths = [filename:join(Path, F) || F <- More],
+                    NewPaths = lists:flatmap(fun dir_files/1, Paths),
+                    [Path | NewPaths];
+                _ ->
+                    [Path]
+            end;
+        _ ->
             [Path]
     end.
 
-relative_path(AbsFile, AbsDir) ->
-    rebar_dir:make_relative_path(AbsFile, AbsDir).
-
 update_app_src(App, Version) ->
     AppSrcFile = rebar_app_info:app_file_src(App),
     AppSrc = rebar_file_utils:try_consult(AppSrcFile),
     [{application, Name, Details}] = AppSrc,
     NewDetails = lists:keyreplace(vsn, 1, Details, {vsn, rebar_utils:to_list(Version)}),
     {application, Name, NewDetails}.
+
+relative_to(Path, From) ->
```

Comment:
I tried keeping `rebar_dir:make_relative_path/1` but ran into problems. Can look into that again, but I figured eh, just take the implementation from Elixir's Path since the algo is based on what's going on in hex anyway.